src/core/image_resizer.js

```diff
@@ -0,0 +1,320 @@
+/* Copyright 2023 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FeatureTest, ImageKind, shadow } from "../shared/util.js";
+
+const MIN_IMAGE_DIM = 2048;
+const MAX_IMAGE_DIM = 65536;
+
+// Large images are encoded in using the BMP format (it's a way faster than
+// encoding in PNG because there are no checksums to compute).
+// Unfortunately, createImageBitmap will create a task in the main thread in
+// order to build the bitmap, hence the main thread is blocked during the
+// decoding which can be a bit long (see bug 1817244).
+// When bug 1759728 is fixed we can remove the BMP workaround and at least it
+// should be a way faster to create the bitmap.
```

Comment:
Given that https://bugzilla.mozilla.org/show_bug.cgi?id=1759728 has patches now, it might make sense to wait for that to land before moving forward with this PR?

Reply:
It will likely take a while, and maybe not happen at all, as that change is a bit controversial.

Reply:
As far as I know, even with the mentioned bug fixed we'll still block the main thread so it won't be ideal either.
As I mentioned in my first comment, it's really not ideal but a way better than a blank page.
Here is a profile with the pdf I added in the test suite:
https://share.firefox.dev/3SJoj5k

We spend around half a second in guessing the max-dims/area: it's improvable for Firefox in getting them from prefs.
We spend around 1.3s to decode the image and move it to a BMP, 1.1s to decode the BMP in the main thread and 0.9s to resize it. And I think it's acceptable (at least compared to the current situation).

We can potentially see some new issues when zooming: the images will be large enough to not fit the 2Mpix limit and consequently they'll be removed once the rendering is done which induce to get it from the worker on each redraw. We could maybe have a different strategy here for large images:
 - we cache them for 5/10 s,
 - once an image expires, we can move  it on a canvas (with `transferFromImageBitmap`), then export the canvas to a png blob (which should really reduce the memory use), and keep the blob for 20/30s. We'll have some overhead when drawing the png but nothing compared to recreate the image from the worker.

I think we could resize it in the worker in tiling it, then rescale each tile and draw them in a smaller canvas but we need to figure out how it works at the tiles junctions (we could probably get a slightly larger tile and then recover the extra pixels with the ones of the next tile).

They're different improvements we can do in the future but this patch is just a version 1.0 and really it improves the current situation.

Reply:
> We can potentially see some new issues when zooming: the images will be large enough to not fit the 2Mpix limit and consequently they'll be removed once the rendering is done which induce to get it from the worker on each redraw. We could maybe have a different strategy here for large images:

One thing that I've been wondering, for some time, is if we should perhaps consider simply increasing that limit a bit?

That value originated in PR #2317, and in the decade that have passed the amount of RAM available in (most) devices should have increased a fair bit. (And we also do a better job when it comes to detecting repeated images at both the page and document level, which also helps.)

Another thing we could potentially try is to instead have a total maximum size *per page*, rather than the current image-specific limit?

Anyway, I'd agree that all of this seems better suited for possible follow-up work.

Reply:
I began to have a look on tiling the original image and rescale each tiles and overall it's slightly slower but at least it doesn't block the main thread. The only concern I've is about the quality on the tile junction.
Anyway I'd prefer for now having this solution blocking the main thread but where the quality of the reduction is good.
I think we could add more reftests and then we could try the tiling strategy in a v2 and we could at least compare the quality of the reduction based on the reftests (if they show something wrong).

Reply:
And it's worth noticing that it's overall slower and I only test in supposing the width lower than 32767 which allows to take an image chunk as a pixel array slice. If we have to chunkify along the width we'll have a lot of extra memcpies.